File: dotnet/src/Pactman.NonprofitCheckPlus/Configuration/ClientConfig.cs
using System;
using System.Collections.Generic;

namespace Pactman.NonprofitCheckPlus.Configuration
{
    /// <summary>
    /// Fully-resolved client configuration.
    /// </summary>
    /// <remarks>
    /// The API key is deliberately not a property of this object, so no diagnostic that
    /// reaches for the configuration can print the credential.
    /// </remarks>
    public sealed class ClientConfig
    {
        /// <summary>Default timeout per attempt.</summary>
        public static readonly TimeSpan DefaultTimeout = TimeSpan.FromSeconds(30);

        /// <summary>Initializes the resolved configuration.</summary>
        /// <param name="baseUrl">The host every request is sent to.</param>
        /// <param name="environment">The named environment, or <see langword="null"/> when an explicit base URL was given.</param>
        /// <param name="timeout">Timeout per attempt.</param>
        /// <param name="retry">Retry policy.</param>
        /// <param name="maxRequestsPerSecond">Optional client-side outbound ceiling.</param>
        /// <param name="defaultHeaders">Extra headers sent with every request.</param>
        /// <param name="userAgent">The <c>User-Agent</c> this client reports.</param>
        public ClientConfig(
            string baseUrl,
            PactmanEnvironment? environment,
            TimeSpan timeout,
            RetryOptions retry,
            double? maxRequestsPerSecond,
            IReadOnlyDictionary<string, string> defaultHeaders,
            string userAgent)
        {
            BaseUrl = baseUrl;
            Environment = environment;
            Timeout = timeout;
            Retry = retry;
            MaxRequestsPerSecond = maxRequestsPerSecond;
            DefaultHeaders = defaultHeaders;
            UserAgent = userAgent;
        }

        /// <summary>The host every request is sent to.</summary>
        public string BaseUrl { get; }

        /// <summary>The named environment, or <see langword="null"/> when an explicit base URL was given.</summary>
        public PactmanEnvironment? Environment { get; }

        /// <summary>Timeout per attempt.</summary>
        public TimeSpan Timeout { get; }

        /// <summary>Retry policy.</summary>
        public RetryOptions Retry { get; }

        /// <summary>Optional client-side ceiling on outbound requests per second.</summary>
        public double? MaxRequestsPerSecond { get; }

        /// <summary>Extra headers sent with every request.</summary>
        public IReadOnlyDictionary<string, string> DefaultHeaders { get; }

        /// <summary>The <c>User-Agent</c> this client reports.</summary>
        public string UserAgent { get; }

        /// <summary>A serializable view of the configuration. Never contains the API key.</summary>
        /// <returns>The configuration's fields, keyed by name.</returns>
        public IReadOnlyDictionary<string, object?> ToDictionary() =>
            new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["baseUrl"] = BaseUrl,
                ["environment"] = Environment?.Name(),
                ["timeout"] = Timeout.TotalSeconds,
                ["retry"] = Retry.ToDictionary(),
                ["maxRequestsPerSecond"] = MaxRequestsPerSecond,
                ["userAgent"] = UserAgent,
            };
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/config/ClientConfig.java
package org.pactman.nonprofitcheckplus.config;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;
import org.pactman.nonprofitcheckplus.PactmanEnvironment;
import org.pactman.nonprofitcheckplus.http.HttpExchange;

/**
 * Fully-resolved configuration.
 *
 * <p>The API key is deliberately not here. It is held by the transport and
 * written into the {@code Authorization} header at send time, so no diagnostic
 * view of the configuration can leak it.
 */
public final class ClientConfig {

    private final String baseUrl;
    private final PactmanEnvironment environment;
    private final long timeoutMs;
    private final RetryOptions retry;
    private final Double maxRequestsPerSecond;
    private final Map<String, String> defaultHeaders;
    private final String userAgent;
    private final HttpExchange httpExchange;

    /**
     * Creates the resolved configuration.
     *
     * @param baseUrl              the base URL every request is sent to.
     * @param environment          the named environment, or {@code null} when an
     *                             explicit base URL was given.
     * @param timeoutMs            the overall timeout per attempt.
     * @param retry                the retry policy.
     * @param maxRequestsPerSecond the client-side rate ceiling, or {@code null} when off.
     * @param defaultHeaders       headers sent with every request.
     * @param userAgent            the {@code User-Agent} this client sends.
     * @param httpExchange         the HTTP layer to send through.
     */
    public ClientConfig(
            String baseUrl,
            PactmanEnvironment environment,
            long timeoutMs,
            RetryOptions retry,
            Double maxRequestsPerSecond,
            Map<String, String> defaultHeaders,
            String userAgent,
            HttpExchange httpExchange) {
        this.baseUrl = baseUrl;
        this.environment = environment;
        this.timeoutMs = timeoutMs;
        this.retry = retry;
        this.maxRequestsPerSecond = maxRequestsPerSecond;
        this.defaultHeaders = Collections.unmodifiableMap(
                new LinkedHashMap<>(defaultHeaders == null
                        ? Collections.<String, String>emptyMap()
                        : defaultHeaders));
        this.userAgent = userAgent;
        this.httpExchange = httpExchange;
    }

    /**
     * The base URL every request is sent to.
     *
     * @return an absolute URL with no trailing slash.
     */
    public String baseUrl() {
        return baseUrl;
    }

    /**
     * The named environment in use.
     *
     * @return the environment, or {@code null} when an explicit base URL was given.
     */
    public PactmanEnvironment environment() {
        return environment;
    }

    /**
     * The overall timeout per attempt.
     *
     * @return the timeout in milliseconds.
     */
    public long timeoutMs() {
        return timeoutMs;
    }

    /**
     * The resolved retry policy.
     *
     * @return the policy.
     */
    public RetryOptions retry() {
        return retry;
    }

    /**
     * The client-side ceiling on outbound requests per second.
     *
     * @return the ceiling, or {@code null} when off.
     */
    public Double maxRequestsPerSecond() {
        return maxRequestsPerSecond;
    }

    /**
     * Headers sent with every request.
     *
     * @return the headers.
     */
    public Map<String, String> defaultHeaders() {
        return defaultHeaders;
    }

    /**
     * The {@code User-Agent} this client sends.
     *
     * @return the user agent string.
     */
    public String userAgent() {
        return userAgent;
    }

    /**
     * The HTTP layer requests are sent through.
     *
     * @return the exchange.
     */
    public HttpExchange httpExchange() {
        return httpExchange;
    }

    /**
     * A redacted view of this configuration.
     *
     * <p>The API key is not a property of this object and never appears here.
     *
     * @return the settings, in a stable order.
     */
    public Map<String, Object> toMap() {
        Map<String, Object> fields = new LinkedHashMap<>();
        fields.put("baseUrl", baseUrl);
        fields.put("environment", environment == null ? null : environment.environmentName());
        fields.put("timeoutMs", timeoutMs);
        fields.put("maxRetries", retry.maxRetries());
        fields.put("maxRequestsPerSecond", maxRequestsPerSecond);
        fields.put("userAgent", userAgent);
        fields.put("apiKey", "[redacted]");

        return fields;
    }
}
